include/deal.II/base/mutable_bind.h

```diff
@@ -0,0 +1,320 @@
+                    const std::unique_ptr<Patterns::PatternBase> &pattern =
+                      Patterns::Tools::Convert<TupleType>::to_pattern());
+
+    /**
+     * Set the arguments to use in @p function, for next time
+     * operator()() is called.
+     */
+    void
+    set_arguments(FunctionArgs... args);
+  };
+
+
+
+  /**
+   * Create a MutableBind object from a function pointer and a list of
+   * arguments.
+   *
+   * An example usage is given by:
+   * @code
+   * void my_function(const int &a, const double &b);
+   *
+   * auto bound = mutable_bind(my_function, 1, 2.0);
+   *
+   * bound(); // will execute my_function(1, 2.0);
+   *
+   * bound.set_parameters(2, 3.0);
+   * bound(); // will execute my_function(2, 3.0);
+   *
+   * bound.parse_parameters("3: 4.0");
+   * bound(); // will execute my_function(3, 4.0);
+   * @endcode
+   *
+   * @authors Luca Heltai, Matthias Maier, 2019.
+   */
+  template <typename ReturnType, class... FunctionArgs>
+  MutableBind<ReturnType, FunctionArgs...>
+  mutable_bind(ReturnType (*function)(FunctionArgs...),
+               typename identity<FunctionArgs>::type... arguments);
+
+  /**
+   * Create a MutableBind object from a function pointer, with uninitialized
+   * arguments.
+   *
+   * An example usage of this function is given by:
+   * @code
+   * void my_function(const int &a, const double &b);
+   *
+   * auto bound = mutable_bind(my_function);
+   *
+   * bound.set_parameters(2, 3.0);
+   * bound(); // will execute my_function(2, 3.0);
+   *
+   * bound.parse_parameters("3: 4.0");
+   * bound(); // will execute my_function(3, 4.0);
+   * @endcode
+   *
+   * Notice that, if you do not call one of the MutableBind::set_parameters()
+   * methods, or the MutableBind::parse_parameters(), the arguments passed to
+   * the function object will be initialized with the values coming from the
+   * arguments defalt constructors.
+   *
+   * @authors Luca Heltai, Matthias Maier, 2019.
+   */
+  template <typename ReturnType, class... FunctionArgs>
+  MutableBind<ReturnType, FunctionArgs...>
+    mutable_bind(ReturnType (*function)(FunctionArgs...));
+
+  /**
+   * Create a MutableBind object from a function object and a list of
+   * arguments.
+   *
+   * An example usage is given by:
+   * @code
+   * std::function<void(const int a, const double)> my_function = ...;
+   *
+   * auto bound = mutable_bind(my_function, 1, 2.0);
+   *
+   * bound(); // will execute my_function(1, 2.0);
+   *
+   * bound.set_parameters(2, 3.0);
+   * bound(); // will execute my_function(2, 3.0);
+   *
+   * bound.parse_parameters("3: 4.0");
+   * bound(); // will execute my_function(3, 4.0);
+   * @endcode
+   *
+   * @authors Luca Heltai, Matthias Maier, 2019.
+   */
+  template <typename ReturnType, class... FunctionArgs>
+  MutableBind<ReturnType, FunctionArgs...>
+  mutable_bind(std::function<ReturnType(FunctionArgs...)>,
+               typename identity<FunctionArgs>::type... arguments);
```

Comment:
Here and in the other places, you could also do the rvalue reference thingy. It's going to lead to a proliferation of functions, though.

It might be easier if you didn't duplicate the code snippet for every function, but only said "Like the previous function, except ..."